components/places/src/bookmark_sync/incoming.rs

```diff
@@ -46,39 +44,59 @@ impl<'a> IncomingApplicator<'a> {
                 BookmarkRecordId::from_payload_id(payload.id).as_guid(),
             )?;
         } else {
-            let item: BookmarkItemRecord = payload.into_record()?;
-            match item {
-                BookmarkItemRecord::Bookmark(b) => self.store_incoming_bookmark(timestamp, b)?,
-                BookmarkItemRecord::Query(q) => self.store_incoming_query(timestamp, q)?,
-                BookmarkItemRecord::Folder(f) => self.store_incoming_folder(timestamp, f)?,
-                BookmarkItemRecord::Livemark(l) => self.store_incoming_livemark(timestamp, l)?,
-                BookmarkItemRecord::Separator(s) => self.store_incoming_sep(timestamp, s)?,
-            }
+            let value: JsonValue = payload.into();
+            match value["type"].as_str() {
+                Some("bookmark") => self.store_incoming_bookmark(timestamp, &value)?,
+                Some("query") => self.store_incoming_query(timestamp, &value)?,
+                Some("folder") => self.store_incoming_folder(timestamp, &value)?,
+                Some("livemark") => self.store_incoming_livemark(timestamp, &value)?,
+                Some("separator") => self.store_incoming_sep(timestamp, &value)?,
+                _ => (),
+            };
         }
         Ok(())
     }
 
-    fn store_incoming_bookmark(&self, modified: ServerTimestamp, b: BookmarkRecord) -> Result<()> {
-        let url = match self.maybe_store_href(b.url.as_ref()) {
-            Ok(url) => (Some(url.into_string())),
+    fn store_incoming_bookmark(&self, modified: ServerTimestamp, b: &JsonValue) -> Result<()> {
+        let mut validity = SyncedBookmarkValidity::Valid;
+
+        let record_id = unpack_id("id", b, &mut validity);
+        let parent_record_id = unpack_optional_id("parentid", b, &mut validity);
+        let date_added = unpack_optional_i64("dateAdded", b, &mut validity);
+        let title = unpack_optional_string("title", b, &mut validity);
+        let keyword = unpack_optional_string("keyword", b, &mut validity);
```

Comment:
Sorry, I should have provided better guidance about which fields need `Replace`, and which ones need `Reupload`...and I didn't actually explain the difference between the two! 😕 Let's change these a bit:

* `record_id` should never be missing. We can return an error and fail the sync if it's missing or not a string.
* `parent_record_id` doesn't need to be either `Reupload` or `Replace`. We can extract it as an `Option<String>` (or `Option<&str>`?), and the merger will take care of flagging the item for reupload because the parent ID is missing.
* `date_added`, `title`, and `keyword` should be `Reupload`, not `Replace`.

In general, `Replace` means "this record is fatally invalid, and there's no way we can apply it". We'll try to replace it with a valid local copy, or, failing that, upload a tombstone. I think the only time we really need it is for invalid URLs.

For the other cases, like a title or keyword that's not a string, or a date that doesn't parse, we can still insert the item, leaving it with an empty title or keyword, or fixed date, which we then `Reupload`.

Reply:
Updated validity handling to use `Reupload` instead of `Replace`, except for URLs.